File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int digit_at_position(long long num, int pos);

int main(void)
{
    // prompt user for card number
    string prompt = "Card number: ";
    long long card = get_long_long("%s", prompt);

    // validate positive non-zero
    if (card <= 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // validate checksum

    int total_num_digits = floor(log10(llabs(card))) + 1;
    int sum = 0;
    int doubled;

    // for every other card digit, starting second from the last
    for (int i = 1; i < total_num_digits; i += 2)
    {
        // double the digit
        doubled = digit_at_position(card, i) * 2;

        // if the product has two digits, add both digits to the sum
        if (doubled >= 10)
        {
            sum += doubled / 10 + doubled % 10;
        }
        // otherwise add the single digit to the sum
        else
        {
            sum += doubled;
        }
    }

    // for the remaining card digits
    for (int i = 0; i < total_num_digits; i += 2)
    {
        // add digit to the sum
        sum += digit_at_position(card, i);
    }

    // reject if the last digit of the sum is not zero
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // validate length & starting digits (type)

    int first_digit = digit_at_position(card, total_num_digits - 1);
    int second_digit = digit_at_position(card, total_num_digits - 2);
    int leading = first_digit * 10 + second_digit;

    if (total_num_digits == 15 && (leading == 34 || leading == 37))
    {
        printf("AMEX\n");
    }
    else if (total_num_digits == 16 && (leading >= 51 && leading <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((total_num_digits == 16 || total_num_digits == 13) && first_digit == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

// returns the (int) digit at a particular position - zero-indexed, but starts from right, e.g, 543210
int digit_at_position(long long num, int pos)
{
    long long int truncated_num = num / (long long int) pow(10, pos);
    return (int) (truncated_num % 10);
}

File: pset6/credit/credit.py
# calculates a credit card checksum

from cs50 import get_int
from math import log10, floor


def main():
    # prompt user for card number
    ccNum = get_int("Card number: ")

    # validate positive non-zero
    if ccNum <= 0:
        print("INVALID")
        return 0

    # validate checksum
    ccList = [int(digit) for digit in str(ccNum)] # probably ineffecient to convert between string, mathematical approach likely better
    everyOtherDigit = ccList[-2::-2] # starting with second to last digit
    everyOtherDigitDoubled = [double(digit) for digit in everyOtherDigit]
    remainingDigits = ccList[-1::-2]
    finalSum = sum(everyOtherDigitDoubled) + sum(remainingDigits)
    if (not finalSum % 10 == 0):
        print("INVALID")
        return 0

    # validate length & starting digits
    numDigits = int(log10(ccNum)) + 1
    leadingDigits = ccList[0] * 10 + ccList[1]
    if numDigits == 15 and (leadingDigits == 34 or leadingDigits == 37):
        print("AMEX")
    elif numDigits == 16 and (leadingDigits >= 51 and leadingDigits <= 55):
        print("MASTERCARD")
    elif (numDigits == 16 or numDigits == 13) and ccList[0] == 4:
        print("VISA")
    else:
        print("INVALID")

    return 0

def double(digit):
    '''doubles the digit and returns the sum of the resulting digits'''
    doubled = digit * 2
    if doubled >= 10:
        return floor(doubled / 10) + doubled % 10
    return doubled


if __name__ == "__main__":
    main()